Blish HUD/Controls/TextInputBase.cs

```diff
@@ -0,0 +1,718 @@
+
+        /// <summary>
+        /// Gets or sets the current index of the cursor within the text.
+        /// </summary>
+        public int CursorIndex {
+            get => _cursorIndex;
+            set {
+                if (SetProperty(ref _cursorIndex, value, true)) {
+                    OnCursorIndexChanged(new ValueEventArgs<int>(value));
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gets the length of the text.
+        /// </summary>
+        public int Length => _text.Length;
+
+        protected bool IsShiftDown => GameService.Input.Keyboard.ActiveModifiers.HasFlag(ModifierKeys.Shift);
+        protected bool IsCtrlDown  => GameService.Input.Keyboard.ActiveModifiers.HasFlag(ModifierKeys.Ctrl);
+
+        protected bool _multiline;
+        protected bool _caretVisible;
+        protected bool _cursorMoved;
+
+        private TimeSpan _lastInvalidate;
+        private bool     _insertMode;
+        private bool     _suppressRedoStackReset;
+
+        private readonly UndoRedoStack _undoStack = new UndoRedoStack();
+        private readonly UndoRedoStack _redoStack = new UndoRedoStack();
+
+        private readonly Dictionary<Keys, KeyRepeatState> _keyRepeatStates;
+
+        public TextInputBase() {
+            _lastInvalidate  = DateTime.MinValue.TimeOfDay;
+            _keyRepeatStates = new Dictionary<Keys, KeyRepeatState>();
+
+            Input.Mouse.LeftMouseButtonReleased += OnGlobalMouseLeftMouseButtonReleased;
+            Input.Keyboard.KeyStateChanged      += OnGlobalKeyboardKeyStateChanged;
+        }
+
+        private void OnTextInput(string value) {
+            bool ctrlDown = this.IsCtrlDown;
+
+            foreach (char c in value) {
+                if (char.IsControl(c)) continue;
+                if (_font.GetCharacterRegion(c) == null) continue;
+
+                if (ctrlDown) {
+                    switch (c) {
+                        case 'c':
+                            HandleCopy();
+                            return;
+                        case 'x':
+                            HandleCut();
+                            return;
+                        case 'v':
+                            HandlePaste();
+                            return;
+                        case 'z':
+                            HandleUndo();
+                            return;
+                        case 'y':
+                            HandleRedo();
+                            return;
+                        case 'a':
+                            SelectAll();
+                            return;
+                    }
+                }
+
+                InputChar(c);
+            }
+        }
+
+        private void DeleteChars(int index, int length) {
+            if (length <= 0) return;
+
+            SetText(_text.Substring(0, index) + _text.Substring(index + length), true);
+        }
+
+        private string RemoveUnsupportedCharacters(string value) {
```

Comment:
This could possibly result in a StackOverflow if value has a lot of different unsupported characters

```suggestion
        private string RemoveUnsupportedCharacters(string value){
    var result = new StringBuilder();
    foreach (char c in value) {
        (_font.GetCharacterRegion(c) != null || _multiline && c == NEWLINE) {
            result.Append(c);
        }
    }

    return result.ToString();
}
```

Reply:
Updated!